store/src/db/errors.rs

```diff
@@ -3,28 +3,89 @@ use std::io;
 use deadpool_sqlite::PoolError;
 use miden_crypto::{merkle::MerkleError, utils::DeserializationError};
 use miden_node_proto::block_header::BlockHeader;
+use prost::DecodeError;
+use rusqlite::types::FromSqlError;
 use thiserror::Error;
+use tokio::sync::oneshot::error::RecvError;
+
+use crate::errors::StateError;
+
+#[derive(Debug, Error)]
+pub enum DbError {
+    #[error("Missing database connection: {0}")]
+    MissingDbConnection(#[from] PoolError),
+    #[error("SQLite error: {0}")]
+    SqliteError(#[from] rusqlite::Error),
+    #[error("SQLite error: {0}")]
+    FromSqlError(#[from] FromSqlError),
+    #[error("SQLite migration error: {0}")]
+    SqliteMigrationError(#[from] rusqlite_migration::Error),
+    #[error("I/O error: {0}")]
+    IoError(#[from] io::Error),
+    #[error("Pool build error: {0}")]
+    PoolBuildError(#[from] deadpool_sqlite::BuildError),
+    #[error("Prost decode error: {0}")]
+    DecodeError(#[from] DecodeError),
+    #[error("SQLite pool interaction task failed: {0}")]
+    InteractionTaskError(#[from] InteractionTaskError),
+    #[error("Genesis block error: {0}")]
+    GenesisBlockError(#[from] GenesisBlockError),
+    #[error("State error: {0}")]
+    StateError(Box<StateError>),
+    #[error("Block database is empty")]
+    BlockDbIsEmpty,
+    #[error("Decoding nullifier from database failed: {0}")]
+    NullifierDecodingError(DeserializationError),
+    #[error("Block applying was broken because of closed channel on state side: {0}")]
+    BlockApplyingBrokenBecauseOfClosedChannel(RecvError),
+}
+
+impl From<StateError> for DbError {
+    fn from(value: StateError) -> Self {
+        Self::StateError(Box::new(value))
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum InteractionTaskError {
+    #[error("Migration task failed: {0}")]
+    MigrationTaskFailed(String),
+    #[error("Select block headers task failed: {0}")]
+    SelectBlockHeadersTaskFailed(String),
+    #[error("Get nullifiers task failed: {0}")]
+    GetNullifiersTaskFailed(String),
+    #[error("Get notes task failed: {0}")]
+    GetNotesTaskFailed(String),
+    #[error("Get accounts task failed: {0}")]
+    GetAccountsTaskFailed(String),
+    #[error("Get block header task failed: {0}")]
+    GetBlockHeaderTaskFailed(String),
+    #[error("Get block headers task failed: {0}")]
+    GetBlockHeadersTaskFailed(String),
+    #[error("Get account hashes task failed: {0}")]
+    GetAccountHashesTaskFailed(String),
+    #[error("Apply block task failed: {0}")]
+    ApplyBlockTaskFailed(String),
+}
```

Comment:
I'm not sure I get the naming for this error and its variants. If I'm understanding correctly, this error happens when we are trying to query the database, right? If so, should we just call it `QueryError`?

As for variants, should we make them consistent with function names? For example: `SelectAccountsError` instead of `GetAccountsTaskFailed`.

Reply:
@bobbinth 
> I'm not sure I get the naming for this error and its variants. If I'm understanding correctly, this error happens when we are trying to query the database, right? If so, should we just call it QueryError?

These errors are used for wrapping errors raised by deadpool's `interact` method. The only two cases of such errors are panic and abortion of user's callback closure.

> As for variants, should we make them consistent with function names? For example: SelectAccountsError instead of GetAccountsTaskFailed.

Done.

Reply:
> These errors are used for wrapping errors raised by deadpool's `interact` method. The only two cases of such errors are panic and abortion of user's callback closure.

What happens if there is error with the query itself (e.g., malformed SQL statement)? For example:
```Rust
pub async fn select_accounts(&self) -> Result<Vec<AccountInfo>> {
    self.pool
        .get()
        .await?
        .interact(sql::select_accounts)
        .await
        .map_err(|err| InteractionTaskError::SelectAccountsTaskFailed(err.to_string()))?
}
```
What would happen if `select_accounts` returns an error? (does it get swallowed somewhere?)

Reply:
No-no, it doesn't get swallowed, it will be returned in `Result` of `select_accounts`. In `sql::select_accounts` this `rusqlite::Error` error will be wrapped with `DbError::SqliteError(err)` and passed through the parent `select_accounts`. In a case of `InteractionTaskError`, it will be also converted to `DbError::InteractionTaskError(err)` (look at question mark in the line with `map_err`) and passed through.

Reply:
Ah - I see. So the value returned from `.interact().await` is `Result<Result<_, DbError>, InteractError>` - right?

Reply:
@bobbinth right!